cmd/admin.go

```diff
@@ -701,7 +707,27 @@ func runUpdateSMTP(c *cli.Context) error {
 
 	source.Cfg = smtpConfig
 
-	return auth_model.UpdateSource(source)
+	return updateSource(ctx, source)
+}
+
+func createSource(ctx context.Context, source *auth_model.Source) error {
+	if err := auth_model.CreateSource(source); err != nil {
+		return err
+	}
+
+	audit.Record(audit.SystemAuthenticationSourceAdd, audit.NewCLIUser(), nil, source, "Created authentication source %s [%s].", source.Name, source.Type.String())
```

Comment:
There's a race condition here that would cause the audit log to be incomplete in some instances, such as if the process crashed right after an event (e.g. creating a source) occurred. Maybe the audit Record step can be split into two steps, a pre-step and a post-step. The pre-step records the action to the audit log, and the post-step confirms that the action completed. That way, even if the process crashes, then audit log reflects that the action was at least attempted if not completed.

Reply:
Also, from what I understand the queue operates in an asynchronous way. This would cause similar susceptibility to audit log incompleteness from what I understand, since an action could be performed, complete, and even an entry added to the queue, but then if the process crashes then that log (as well as other logs) could be lost since they were not yet processed by the queue.

Reply:
Slight revision to my first comment: maybe we can just log the audit event before the action occurs, then have a `rollback` call if it fails.

Reply:
That would require to cache the event until committed. And then it would still be lost if the process crashes. I don't think you can optimize code against program crashes.

Reply:
> maybe we can just log the audit event before the action occurs, then have a rollback call if it fails.

I don't think that is a good idea, the architecture is `add only` at the moment, which sounds a lot better than an architecture that can also delete/withdraw logs.

What we could do instead, which sounds like a much better idea to me, is https://github.com/go-gitea/gitea/pull/24257/files#r1333436722 (perhaps also with a `shutdown` audit log, even though I'm not sure it will always work, for example if Gitea is killed with a `SIGKILL`).
That way, you know that there might be some lost logs exactly before the instance was restarted, especially if there is no `gitea has terminated` before it.
Everything else makes this problem much more complicated for minimal gain.